HW2/P5/driver.py

```diff
@@ -80,3 +96,18 @@ def randcolor():
             # SUBPROBLEM 3: sort objects by location.  Be sure to update the
             # grid if objects' indices change!  Also be sure to sort the
             # velocities with their object positions!
+
+            # get the locations
+            grid_loc = (positions/grid_spacing).astype(int)
+
+            # Morton sort and get objects' indices 
+            mortonsort = map(lambda a,b: cmp_zorder(a,b), grid_loc, np.array(list(grid_loc[1:])+list(grid_loc[:1])))
+            grid_index = np.argsort(mortonsort)
+            velocities, positions= velocities[grid_index], positions[grid_index]
+            # set old positions as -1
+            grid = - np.ones((grid_size, grid_size), dtype=np.uint32) 
+            # set new positions
+            for i in range(num_balls):
```

Comment:
You can vectorize this operation, in the same way as the initialization above.  This might speed things up?